src/monaco/ls.ts

```diff
@@ -0,0 +1,624 @@
+            }
+
+            return modelsMap.value.get(fileName)?.getVersionId().toString() ?? 'unknown version';
+        },
+        getScriptSnapshot(fileName: string): ts.IScriptSnapshot | undefined {
+            console.log('getScriptSnapshot', fileName);
+            let scriptSnapshot = scriptSnapshots.get(fileName);
+            if (!scriptSnapshot || scriptSnapshot.getText(0, scriptSnapshot.getLength()) !== this.readFile(fileName)) {
+                const fileContent = this.readFile(fileName);
+                if (fileContent !== undefined) {
+                    scriptSnapshot = ts.ScriptSnapshot.fromString(fileContent);
+                    scriptSnapshots.set(fileName, scriptSnapshot);
+                }
+            }
+            return scriptSnapshot;
+        },
+        getCurrentDirectory(): string {
+            console.log('getCurrentDirectory');
+            return '/';
+        },
+        getDefaultLibFileName(options: ts.CompilerOptions): string {
+            console.log('getDefaultLibFileName', ts.getDefaultLibFileName(options));
+
+            return ts.getDefaultLibFileName(options);
+        },
+    };
+    const sys: ts.System = {
+        args: [],
+        newLine: '\n',
+        useCaseSensitiveFileNames: false,
+        write(s: string): void {
+            throw new Error('Function not implemented.');
+        },
+        readFile(path: string, encoding?: string): string | undefined {
+            throw new Error('Function not implemented.');
+        },
+        writeFile(path: string, data: string, writeByteOrderMark?: boolean): void {
+            throw new Error('Function not implemented.');
+        },
+        resolvePath(path: string): string {
+            throw new Error('Function not implemented.');
+        },
+        fileExists(path: string): boolean {
+            throw new Error('Function not implemented.');
+        },
+        directoryExists(path: string): boolean {
+            throw new Error('Function not implemented.');
+        },
+        createDirectory(path: string): void {
+            throw new Error('Function not implemented.');
+        },
+        getExecutingFilePath(): string {
+            throw new Error('Function not implemented.');
+        },
+        getCurrentDirectory(): string {
+            throw new Error('Function not implemented.');
+        },
+        getDirectories(path: string): string[] {
+            throw new Error('Function not implemented.');
+        },
+        readDirectory(
+            path: string,
+            extensions?: readonly string[],
+            exclude?: readonly string[],
+            include?: readonly string[],
+            depth?: number,
+        ): string[] {
+            throw new Error('Function not implemented.');
+        },
+        exit(exitCode?: number): void {
+            throw new Error('Function not implemented.');
+        },
+    };
+    const tsWithAny = ts as any;
+    tsWithAny.setSys(sys);
```

Comment:
@johnsoncodehk Hi, It's a bit hack. But I'm not sure where and how in volar or TypeScript are using `ts.sys`....